File: typo-tester-typo-java/src/java/adventureworks/DSLTest.java
package adventureworks;

import adventureworks.customtypes.Defaulted;
import adventureworks.customtypes.TypoLocalDate;
import adventureworks.humanresources.employee.EmployeeRepoImpl;
import adventureworks.person.businessentity.BusinessentityRepoImpl;
import adventureworks.person.emailaddress.EmailaddressRepoImpl;
import adventureworks.person.person.PersonRepoImpl;
import adventureworks.public_.Name;
import adventureworks.sales.salesperson.SalespersonRepoImpl;
import adventureworks.userdefined.FirstName;
import org.junit.Test;

import java.util.Optional;

import static org.junit.Assert.*;

/**
 * Tests for DSL join chains - equivalent to Scala DSLTest.
 */
public class DSLTest extends SnapshotTest {
    private final BusinessentityRepoImpl businessentityRepoImpl = new BusinessentityRepoImpl();
    private final PersonRepoImpl personRepoImpl = new PersonRepoImpl();
    private final EmployeeRepoImpl employeeRepoImpl = new EmployeeRepoImpl();
    private final SalespersonRepoImpl salespersonRepoImpl = new SalespersonRepoImpl();
    private final EmailaddressRepoImpl emailaddressRepoImpl = new EmailaddressRepoImpl();

    @Test
    public void works() {
        WithConnection.run(c -> {
            var testInsert = new TestInsert(new java.util.Random(0), new DomainInsertImpl());

            // Create businessentity
            var businessentityRow = testInsert.personBusinessentity(
                    new Defaulted.UseDefault<>(),
                    new Defaulted.UseDefault<>(),
                    new Defaulted.UseDefault<>(),
                    c
            );

            // Create person
            var personRow = testInsert.personPerson(
                    businessentityRow.businessentityid(),
                    "EM",  // persontype
                    new FirstName("a"),
                    Optional.empty(),  // title
                    Optional.empty(),  // middlename
                    new Name("lastname"),
                    Optional.empty(),  // suffix
                    Optional.empty(),  // additionalcontactinfo
                    Optional.empty(),  // demographics
                    new Defaulted.UseDefault<>(),  // namestyle
                    new Defaulted.UseDefault<>(),  // emailpromotion
                    new Defaulted.UseDefault<>(),  // rowguid
                    new Defaulted.UseDefault<>(),  // modifieddate
                    c
            );

            // Create emailaddress
            testInsert.personEmailaddress(
                    personRow.businessentityid(),
                    Optional.of("a@b.c"),
                    new Defaulted.UseDefault<>(),
                    new Defaulted.UseDefault<>(),
                    new Defaulted.UseDefault<>(),
                    c
            );

            // Create employee
            var employeeRow = testInsert.humanresourcesEmployee(
                    personRow.businessentityid(),
                    new TypoLocalDate(java.time.LocalDate.of(1998, 1, 1)),
                    "M",  // maritalstatus
                    "M",  // gender
                    new TypoLocalDate(java.time.LocalDate.of(1997, 1, 1)),
                    "12345",  // nationalidnumber
                    "adventure-works\\a",  // loginid
                    "Test Job",  // jobtitle
                    new Defaulted.UseDefault<>(),
                    new Defaulted.UseDefault<>(),
                    new Defaulted.UseDefault<>(),
                    new Defaulted.UseDefault<>(),
                    new Defaulted.UseDefault<>(),
                    new Defaulted.UseDefault<>(),
                    new Defaulted.UseDefault<>(),
                    c
            );

            // Create salesperson
            var salespersonRow = testInsert.salesSalesperson(
                    employeeRow.businessentityid(),
                    Optional.empty(),  // territoryid
                    Optional.empty(),  // salesquota
                    new Defaulted.UseDefault<>(),
                    new Defaulted.UseDefault<>(),
                    new Defaulted.UseDefault<>(),
                    new Defaulted.UseDefault<>(),
                    new Defaulted.UseDefault<>(),
                    new Defaulted.UseDefault<>(),
                    c
            );

            // Test join chain: salesperson -> employee -> person -> businessentity, then join with email
            var query = salespersonRepoImpl.select()
                    .where(sp -> sp.rowguid().isEqual(salespersonRow.rowguid()))
                    .joinFk(sp -> sp.fkHumanresourcesEmployee(), employeeRepoImpl.select())
                    .joinFk(sp_e -> sp_e._2().fkPersonPerson(), personRepoImpl.select())
                    .joinFk(sp_e_p -> sp_e_p._2().fkBusinessentity(), businessentityRepoImpl.select())
                    .join(emailaddressRepoImpl.select().orderBy(e -> e.rowguid().asc()))
                    .on(sp_e_p_b_email -> sp_e_p_b_email._2().businessentityid().isEqual(sp_e_p_b_email._1()._2().businessentityid()));

            // Self-join the query
            var doubled = query
                    .join(query)
                    .on(left_right -> left_right._1()._1()._1()._2().businessentityid().isEqual(left_right._2()._1()._1()._2().businessentityid()));

            // Add snapshot test - same as Scala version
            compareFragment("doubled", doubled.sql());

            var results = doubled.toList(c);
            results.forEach(System.out::println);
            assertEquals(1, doubled.count(c));
        });
    }
}

File: typo-tester-typo-kotlin/src/test/kotlin/adventureworks/DSLTest.kt
package adventureworks

import adventureworks.customtypes.TypoLocalDate
import adventureworks.humanresources.employee.EmployeeRepoImpl
import adventureworks.person.businessentity.BusinessentityRepoImpl
import adventureworks.person.emailaddress.EmailaddressRepoImpl
import adventureworks.person.person.PersonRepoImpl
import adventureworks.sales.salesperson.SalespersonRepoImpl
import adventureworks.userdefined.FirstName
import org.junit.Assert.assertEquals
import org.junit.Test
import kotlin.random.Random

class DSLTest {
    private val businessentityRepoImpl = BusinessentityRepoImpl()
    private val personRepoImpl = PersonRepoImpl()
    private val employeeRepoImpl = EmployeeRepoImpl()
    private val salespersonRepoImpl = SalespersonRepoImpl()
    private val emailaddressRepoImpl = EmailaddressRepoImpl()

    @Test
    fun works() {
        WithConnection.run { c ->
            val testInsert = TestInsert(Random(0), DomainInsert)
            val businessentityRow = testInsert.personBusinessentity(c)
            val personRow = testInsert.personPerson(c, businessentityRow.businessentityid, persontype = "EM", firstname = FirstName("a"))
            testInsert.personEmailaddress(c, personRow.businessentityid, emailaddress = "a@b.c")
            val employeeRow = testInsert.humanresourcesEmployee(
                c,
                personRow.businessentityid,
                gender = "M",
                maritalstatus = "M",
                birthdate = TypoLocalDate("1998-01-01"),
                hiredate = TypoLocalDate("1997-01-01")
            )
            val salespersonRow = testInsert.salesSalesperson(c, employeeRow.businessentityid)

            val q = salespersonRepoImpl.select()
                .where { it.rowguid().isEqual(salespersonRow.rowguid) }
                .joinFk({ it.fkHumanresourcesEmployee() }, employeeRepoImpl.select())
                .joinFk({ it._2().fkPersonPerson() }, personRepoImpl.select())
                .joinFk({ it._2().fkBusinessentity() }, businessentityRepoImpl.select())
                .join(emailaddressRepoImpl.select().orderBy { it.rowguid().asc() })
                .on { sp_e_p_b_email -> sp_e_p_b_email._2().businessentityid().isEqual(sp_e_p_b_email._1()._2().businessentityid()) }
                .joinOn(salespersonRepoImpl.select()) { sp_e_p_b_email_s2 ->
                    sp_e_p_b_email_s2._1()._1()._1()._2().businessentityid().underlying.isEqual(sp_e_p_b_email_s2._2().businessentityid().underlying)
                }

            val doubled = q.join(q).on { left_right ->
                left_right._1()._1()._1()._1()._2().businessentityid().isEqual(left_right._2()._1()._1()._1()._2().businessentityid())
            }

            doubled.toList(c).forEach { println(it) }
            val count = doubled.count(c)
            assertEquals(1, count.toInt())

            SnapshotTest.compareFragment("DSLTest", "doubled", doubled.sql())
        }
    }
}
